tests/platform_tests/api/test_psu.py

```diff
@@ -0,0 +1,108 @@
+
+    num_psus = None
+
+    @pytest.fixture(scope="function", autouse=True)
+    def setup(self, platform_api_conn):
+        if self.num_psus is None:
+            try:
+                self.num_psus = int(chassis.get_num_psus(platform_api_conn))
+            except:
+                pytest.fail("num_psus is not an integer")
+
+
+    def test_fans(self, duthost, localhost, platform_api_conn):
+        ''' PSU fan test '''
+        for psu_id in range(self.num_psus):
+            try:
+                num_fans = int(psu.get_num_fans(platform_api_conn, psu_id))
+            except:
+                pytest.fail("num_fans is not an integer!")
+
+            fan_list = psu.get_all_fans(platform_api_conn, psu_id)
+            if self.expect(fan_list is not None, "Failed to retrieve fans of PSU {}".format(psu_id)):
+                self.expect(isinstance(fan_list, list) and len(fan_list) == num_fans, "Fans of PSU {} appear to be incorrect".format(psu_id))
+
+            for i in range(num_fans):
+                fan = psu.get_fan(platform_api_conn, psu_id, i)
+                self.expect(fan and fan == fan_list[i], "Fan {} of PSU {} is incorrect".format(i, psu_id))
+        self.assert_expectations()
+
+
+    def test_power(self, duthost, localhost, platform_api_conn):
+        ''' PSU power test '''
+        for psu_id in range(self.num_psus):
+            voltage = psu.get_voltage(platform_api_conn, psu_id)
+            current = psu.get_current(platform_api_conn, psu_id)
+            power = psu.get_power(platform_api_conn, psu_id)
+            self.expect(abs(power - (voltage*current)) < power*0.1, "PSU {} reading does not make sense \
+                (power:{}, voltage:{}, current:{})".format(psu_id, power, voltage, current))
+
+            powergood_status = psu.get_powergood_status(platform_api_conn, psu_id)
+            if self.expect(powergood_status is not None, "Failed to retrieve operational status of PSU {}".format(psu_id)):
+                self.expect(powergood_status is True, "PSU {} is not operational".format(psu_id))
+
+            high_threshold = psu.get_voltage_high_threshold(platform_api_conn, psu_id)
+            low_threshold = psu.get_voltage_low_threshold(platform_api_conn, psu_id)
+            if self.expect(high_threshold is not None and low_threshold is not None, "Failed to retrieve the voltage threshold values of PSU {}".format(psu_id)):
+                self.expect(voltage < high_threshold and voltage > low_threshold, "Voltage {} of PSU {} is not in between {} and {}".format(voltage, psu_id, low_threshold, high_threshold))
+        self.assert_expectations()
+
+
+    def test_temperature(self, duthost, localhost, platform_api_conn):
+        ''' PSU temperature test '''
+        for psu_id in range(self.num_psus):
+            temperature = psu.get_temperature(platform_api_conn, psu_id)
+            self.expect(temperature is not None, "Failed to retrieve temperature of PSU {}".format(psu_id))
```

Comment:
Suggest to also check that value if of the expected type.

Reply:
@jleveque the value validation check is down below after reading the temperature threshold.

Reply:
There you're comparing `temperature < temp_threshold`. However, we should also ensure we get the proper type, as Python could allow the vendor to return an int instead of a float, etc, which the comparison alone won't catch. I have made similar checks in the SFP tests (example: https://github.com/Azure/sonic-mgmt/blob/master/tests/platform_tests/api/test_sfp.py#L223).

I think we should always make three checks on return values;
1. Value is not None (i.e., vendor has implemented the API call)
2. Value is of appropriate type (float, int, bool, str, etc.)
3. Value is acceptable (within range, tolerance, etc.), for values we have data to compare against